src/codegen/mod.rs

```diff
@@ -3724,40 +3722,62 @@ impl CodeGenerator for ObjCInterface {
             let ambiquity =
                 instance_method_names.contains(&class_method.rust_name());
             let prefix = if ambiquity { "class_" } else { "" };
-            let (impl_item, trait_item) = objc_method_codegen(
+            let (impl_item, _trait_item) = objc_method_codegen(
                 ctx,
                 class_method,
                 Some(self.name()),
                 prefix,
             );
             impl_items.push(impl_item);
-            trait_items.push(trait_item)
         }
 
         let trait_name = ctx.rust_ident(self.rust_name());
-
+        let trait_constraints = quote! {
+            Sized + std::ops::Deref + objc::Message
+        };
         let trait_block = if self.is_template() {
             let template_names: Vec<Ident> = self
                 .template_names
                 .iter()
                 .map(|g| ctx.rust_ident(g))
                 .collect();
+
             quote! {
-                pub trait #trait_name <#(#template_names),*>{
-                    #( #trait_items )*
+                pub trait #trait_name <#(#template_names),*> : #trait_constraints {
+                    #( #impl_items )*
                 }
             }
         } else {
             quote! {
-                pub trait #trait_name {
-                    #( #trait_items )*
+                pub trait #trait_name : #trait_constraints {
+                    #( #impl_items )*
                 }
             }
         };
 
-        let ty_for_impl = quote! {
-            id
-        };
+        let struct_name = ctx.rust_ident(self.struct_name());
+        if !(self.is_category() || self.is_protocol()) {
+            let struct_block = quote! {
+                pub struct #struct_name(id);
+                impl std::ops::Deref for #struct_name {
+                    type Target = id;
+                    fn deref(&self) -> &Self::Target {
+                        unsafe { ::core::mem::transmute(self.0) }
```

Comment:
I don't get this, this doesn't seem right. You're turning an `id` into a `&id`. Why? And why can't this just be `&self.0`, what am I missing?

Reply:
Oh, that's much better. I think I tried that but had issues with some of the trait constraints so I defaulted to the suggestion in #109.